sql-plugin/src/main/java/com/nvidia/spark/rapids/InternalRowToColumnarBatchIterator.java

```diff
@@ -110,73 +119,103 @@ public ColumnarBatch next() {
+        if (retryCount > 0) {
+          // In case we hit the corner case, and we had to increase the dataLength, then we need
+          // to restore the original capacity so that we do not carry on with large memory buffers.
+          numRowsEstimate = initialNumRowsEstimate;
+          dataLength = initialDataLength;
+        }
+
+        if (numInputRows != null) {
+          numInputRows.add(currentRow);
+        }
+        if (numOutputRows != null) {
+          numOutputRows.add(currentRow);
         }
-        buildRange = NvtxWithMetrics.apply("RowToColumnar: build", NvtxColor.GREEN, Option.apply(opTime));
-        devColumn = hostColumn.copyToDevice();
+        if (numOutputBatches != null) {
+          numOutputBatches.add(1);
+        }
+        // Now that we have filled the buffers with the data, we need to turn them into a
+        // HostColumnVector and copy them to the device so the GPU can turn it into a Table.
+        // To do this we first need to make a HostColumnCoreVector for the data, and then
+        // put that into a HostColumnVector as its child.  This the basics of building up
+        // a column of lists of bytes in CUDF but it is typically hidden behind the higher level
+        // APIs.
+        dataBuffer.incRefCount();
+        offsetsBuffer.incRefCount();
+        try (HostColumnVectorCore dataCv =
+                 new HostColumnVectorCore(DType.INT8, dataOffset, Optional.of(0L),
+                     dataBuffer, null, null, new ArrayList<>());
+             HostColumnVector hostColumn = new HostColumnVector(DType.LIST,
+                 currentRow, Optional.of(0L), null, null,
+                 offsetsBuffer, Collections.singletonList(dataCv))) {
+
+          long ct = System.nanoTime() - collectStart;
+          streamTime.add(ct);
+
+          // Grab the semaphore because we are about to put data onto the GPU.
+          TaskContext tc = TaskContext.get();
+          if (tc != null) {
+            GpuSemaphore$.MODULE$.acquireIfNecessary(tc, semaphoreWaitTime);
+          }
+          buildRange = NvtxWithMetrics.apply("RowToColumnar: build", NvtxColor.GREEN, Option.apply(opTime));
+          devColumn = hostColumn.copyToDevice();
+        }
+      } catch (BufferOverflowException ex) {
+        // Handle corner case when the dataLength is too small to copy a single row.
+        // We retry after doubling the bufferSize.
+        // dataLength can be considered a rough estimate of a single row.
+        if (dataLength >= JCudfUtil.JCUDF_MAX_ROW_BUFFER_LENGTH) {
+          // proceed with throwing exception.
+          throw new RuntimeException(
+              "JCudf row is too large to fit in MemoryBuffer", ex);
+        }
+        long newRowSizeEst = Math.min(dataLength << 1, JCudfUtil.JCUDF_MAX_ROW_BUFFER_LENGTH);
+        // Recalculate the dataLength based on the new size estimate
+        setBuffersCapacities(newRowSizeEst, goalTargetSize);
+        retryCount++;
```

Comment:
I had considered that scenario.

- Scenario1: If fillBatch throws an exception, it won't be handled by the catch clause. It will exits the loop.
- Scenario2: we fail to copy at least one row, then `BufferOverflowException` will be handled by the catch clause.
  - rowSizeEstimate is an integer aligned to 8. So, It cannot exceed `JCUDF_MAX_ROW_BUFFER_LENGTH`.
  - `dataLength` is incremented by left-shift operator. So it doubles every iteration. This means that there is no delta-error that causes the step-increment to stall.
  - `dataLength` represents multiple rows of `rowSizeEstimate`. Therefore, it has to be `dataLength >= rowSizeEstimate`
  - The only way `dataLength` would not increase is there is an overflow which is not likely to happen in that flow.